### PR TITLE
feat: minimap and full screen map (disabled by default)

### LINKED ISSUE #64: [Game UI] Minimap + Full-Screen Map
Implement full-screen map with fullscreen canvas.

- X on top left to close
- display only top-most block on the minimap
- minimap should preserve position after reopen when its changed from the player position
- display entities 

API:
Get the color of each block:
```ts
// Player position:
bot.entity.position
// Get block name:
bot.world.getBlock(new Vec3(x, y, z))?.name // e.g. oak_log
// Skip blocks that have boundingBox === 'empty' and don't have the color (for example lava has the color so it should be rendered)

// You can get block colors from prismarine-viewer/viewer/lib/moreBlockDataGenerated.json (just import this file)
```

And there is a json file (fallback color - black).

Minimap in HUD should be small and look like this:

image

### PR DESCRIPTION
**Pull request by gguio**
fixes #64

[LAST GREENFIELD DEMO](https://minimap.mcraft.fun/?mapDir=https%3A%2F%2Fbucket.mcraft.fun%2FGreenfield+v0.5.1%2Fmap-index.json)
[LAST SINGLEPLAYER DEMO](https://minimap.mcraft.fun/?singleplayer=1&version=1.20.4)
- [x] add warp window changes & quick to
- [x] render current player pos on map
- [ ] Render all chunks in single player worlds with all perf optimisations

### CODE REVIEW

File: src/react/MinimapProvider.tsx [hunk 1]
```diff
@@ -15,19 +16,29 @@ export class DrawerAdapterImpl extends TypedEventEmitter<MapUpdates> implements
...
   world: string
   currChunk: PCChunk | undefined
   currChunkPos: { x: number, z: number } = { x: 0, z: 0 }
 
-  constructor (pos?: Vec3, initialWarps?: WorldWarp[]) {
+  constructor (pos?: Vec3) {
     super()
     this.playerPosition = pos ?? new Vec3(0, 0, 0)
     this.warps = warps
+    if (localServer) {
```
Comment (gguio) [inline src/react/MinimapProvider.tsx:73]: this subscribe doesnt fire

File: src/react/Input.tsx [hunk 2]
```diff
@@ -9,10 +9,10 @@ interface Props extends React.ComponentProps<'input'> {
   validateInput?: (value: string) => CSSProperties | undefined
 }
 
-export default ({ autoFocus, rootStyles, inputRef, validateInput, ...inputProps }: Props) => {
+export default ({ autoFocus, rootStyles, inputRef, validateInput, defaultValue, ...inputProps }: Props) => {
   const ref = useRef<HTMLInputElement>(null!)
   const [validationStyle, setValidationStyle] = useState<CSSProperties>({})
-  const [value, setValue] = useState(inputProps.value ?? '')
+  const [value, setValue] = useState(defaultValue ?? '')
```
Comment (zardoy) [inline src/react/Input.tsx:15]:
```suggestion
  const [value, setValue] = useState(defaultValue ?? props.value ?? '')
```

**Review by zardoy — CHANGES_REQUESTED** [review]

File: src/react/MinimapProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+        const [x, z] = this.loadingChunksQueue.values().next().value.split(',').map(Number)
+        this.loadChunk(x, z)
+        this.loadingChunksQueue.delete(`${x},${z}`)
+      })
+    } else {
+      const storageWarps = localStorage.getItem(`warps: ${loadedGameState.username} ${loadedGameState.serverIp ?? ''}`)
+      this.overwriteWarps(JSON.parse(storageWarps ?? '[]'))
+    }
+    this.isOldVersion = versionToNumber(bot.version) < versionToNumber('1.13')
+    console.log('version is old:', this.isOldVersion)
```
Comment (zardoy) [inline src/react/MinimapProvider.tsx]: Useless Logging

**Review by zardoy — CHANGES_REQUESTED** [review]

**Comment by Phoenix616** [comment]
This looks pretty cool, is there a way to pre-configure whether its enabled or not? (e.g. via a get parameter in the URL) I took a look at the code and didn't find anything in that direction.

**Comment by zardoy** [comment]
> This looks pretty cool, is there a way to pre-configure whether its enabled or not? (e.g. via a get parameter in the URL) I took a look at the code and didn't find anything in that direction.

@Phoenix616 Yes, its possible to configure almost **every aspect of the application** via parameters in the URL. Remember that you can change the setting of the application in that way: `?setting=<setting_name>:<setting_value>`. So there is an example link with minimap disabled: https://minimap.mcraft.fun/?mapDir=https%3A%2F%2Fbucket.mcraft.fun%2FGreenfield+v0.5.1%2Fmap-index.json&setting=showMinimap:%22never%22 (`&setting=showMinimap:"never"`)

@gguio

**Comment by zardoy** [comment]
disabled minimap by default because of some issues on some platforms & singleplayer maps.

@Phoenix616 would you be interested to test it? the link is minimap.mcraft.fun (just set minimap setting to always)

**Pull request merged**